### DISCUSSION TITLE
Timing the async state machine overhead in .NET 6

### DISCUSSION THREAD

**Discussion by noseratio** [dotnet/runtime#61542] [Q&A]
I've attempted to time the overhead of a simple async state machine method, vs a similar non-async `ContinueWith`-based method, [full source](https://github.com/noseratio/await-vs-ContinueWith/blob/main/Program.cs).

```csharp
static async Task<DBNull> DoSomething()
{
    await Task.Yield();
    return DBNull.Value; 
}

static async Task<DBNull?> TestWithAsyncStateMachine()
{
    try
    {
        return await DoSomething();
    }
    catch
    {
        return null;
    }
}

static Task<DBNull?> TestWithoutAsyncStateMachine()
{
    return DoSomething().ContinueWith(
        t => t.IsCompletedSuccessfully ? t.GetAwaiter().GetResult() : null,
        cancellationToken: default,
        continuationOptions: TaskContinuationOptions.ExecuteSynchronously,
        scheduler: TaskScheduler.Default);
}
```

On a tight loop with 10,000,000 iterations, the difference is nearly non-existent with single-threaded asynchrony (a thread with 
@stephentoub's `AsyncPump`-like [event loop](https://github.com/noseratio/await-vs-ContinueWith/blob/c6ac6a3f428178e0e8e63a18545aa50ed92bc6cc/Program.cs#L85)); that's expected.

Doing the same with multi-threaded asynchrony (where `Task.​Run` is used for scheduling on thread pool) shows ~1-2s difference, even with a fixed number of pool threads and forced garbage collection.  It's still a marginal difference, but I wonder why switching threads is more costly with `async` here. Or whether I have flaws in the timing logic.

To give it a bit more context, here's [the Twiter thread](https://twitter.com/mattleibow/status/1458907669268418564?s=20) that prompted me to do this test. 

<hr>

**Edited**, I've increased the number of iterations to 25,000,000 and the number of worker pool threads to 2000, and I now see nearly the same async overhead across single-threaded and multi-threaded tests, which is what I'd expect.

**Comment by timcassell** [comment]
You should use [BenchmarkDotNet](https://github.com/dotnet/BenchmarkDotNet).

Context switching is expensive. Whether that's switching to a separate thread or posting to the task scheduler. You gave it the option `continuationOptions: TaskContinuationOptions.ExecuteSynchronously` which I would expect it to skip context switching altogether and just execute synchronously, but you also gave it `scheduler: TaskScheduler.Default` which I believe overrides the `ExecuteSynchronously` option. You should either tell it to execute synchronously, or give it a scheduler, not both.

**Reply by stephentoub** [reply]
> you also gave it scheduler: TaskScheduler.Default which I believe overrides the ExecuteSynchronously option. You should either tell it to execute synchronously, or give it a scheduler, not both.

ExecuteSynchronously causes the system to use the scheduler's TryExecuteTaskInline. Specifying TaskScheduler.Default with the ExecuteSynchronously option is fine.

**Reply by noseratio** [reply]
@timcassell thanks for pointing out to BenchmarkDotNet, #TIL! Will give it a try.

> You should either tell it to execute synchronously, or give it a scheduler, not both.

Adding my 2c to Stephen's comment above, I don't think [`ContinueWith`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.continuewith?view=net-5.0) has an override that accepts `continuationOptions`, `cancellationToken` but not `scheduler`, so I had to pass something for `scheduler`.

**Reply by timcassell** [reply]
Well you're just passing a default token anyway, so you could ignore both and just pass the option like the guy in that Twitter thread did. But I guess it doesn't matter either way...
